src/Cli/dotnet/commands/dotnet-run/RunCommand.cs

```diff
@@ -54,133 +75,123 @@ public int Execute()
             }
             catch (InvalidProjectFileException e)
             {
                 throw new GracefulException(
-                    string.Format(LocalizableStrings.RunCommandSpecifiecFileIsNotAValidProject, Project),
+                    string.Format(LocalizableStrings.RunCommandSpecifiecFileIsNotAValidProject, ProjectFileFullPath),
                     e);
             }
         }
 
-        public RunCommand(string configuration,
-            string framework,
-            string runtime,
-            bool noBuild,
-            string project,
-            string launchProfile,
-            bool noLaunchProfile,
-            bool noRestore,
-            bool interactive,
-            IEnumerable<string> restoreArgs,
-            IEnumerable<string> args)
+        private ICommand ApplyLaunchSettingsProfileToCommand(ICommand targetCommand, ProjectLaunchSettingsModel? launchSettings)
         {
-            Configuration = configuration;
-            Framework = framework;
-            Runtime = runtime;
-            NoBuild = noBuild;
-            Project = project;
-            LaunchProfile = launchProfile;
-            NoLaunchProfile = noLaunchProfile;
-            Args = args;
-            RestoreArgs = restoreArgs;
-            NoRestore = noRestore;
-            Interactive = interactive;
+            if (launchSettings != null)
+            {
+                if (!string.IsNullOrEmpty(launchSettings.ApplicationUrl))
+                {
+                    targetCommand.EnvironmentVariable("ASPNETCORE_URLS", launchSettings.ApplicationUrl);
+                }
+
+                targetCommand.EnvironmentVariable("DOTNET_LAUNCH_PROFILE", launchSettings.LaunchProfileName);
+
+                foreach (var entry in launchSettings.EnvironmentVariables)
+                {
+                    string value = Environment.ExpandEnvironmentVariables(entry.Value);
+                    //NOTE: MSBuild variables are not expanded like they are in VS
+                    targetCommand.EnvironmentVariable(entry.Key, value);
+                }
+                if (string.IsNullOrEmpty(targetCommand.CommandArgs) && launchSettings.CommandLineArgs != null)
+                {
+                    targetCommand.SetCommandArgs(launchSettings.CommandLineArgs);
+                }
+            }
+            return targetCommand;
         }
 
-        private bool TryGetLaunchProfileSettingsIfNeeded(out ProjectLaunchSettingsModel launchSettingsModel)
+        private bool TryGetLaunchProfileSettingsIfNeeded(out ProjectLaunchSettingsModel? launchSettingsModel)
         {
             launchSettingsModel = default;
             if (!UseLaunchProfile)
             {
                 return true;
             }
 
-            var buildPathContainer = File.Exists(Project) ? Path.GetDirectoryName(Project) : Project;
-            string propsDirectory;
+            var launchSettingsPath = TryFindLaunchSettings(ProjectFileFullPath);
```

Comment:
Pulled out discovery of launch settings for readability as well. The idea is that we try to find the settings, and then try to find a specific profile inside the settings for later application,